// playwright-smart-reporter/dist/collectors/step-collector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepCollector = void 0;
/**
 * Extracts and processes step timing data from test results
 */
class StepCollector {
    /**
     * Extract step timings from a test result
     * @param result - Playwright TestResult
     * @returns Array of step data with durations and categories
     */
    extractSteps(result) {
        const steps = [];
        // Recursively extract steps from the result
        const processStep = (step) => {
            // Only include meaningful steps (skip internal hooks)
            if (step.category === 'test.step' || step.category === 'pw:api') {
                steps.push({
                    title: step.title,
                    duration: step.duration,
                    category: step.category,
                });
            }
            // Process nested steps
            if (step.steps) {
                for (const nested of step.steps) {
                    processStep(nested);
                }
            }
        };
        for (const step of result.steps) {
            processStep(step);
        }
        // Mark the slowest step if we have any
        if (steps.length > 0) {
            const maxDuration = Math.max(...steps.map((s) => s.duration));
            const slowestIndex = steps.findIndex((s) => s.duration === maxDuration);
            if (slowestIndex !== -1 && maxDuration > 100) {
                steps[slowestIndex].isSlowest = true;
            }
        }
        return steps;
    }
    /**
     * Calculate total duration of all steps
     * @param steps - Array of step data
     * @returns Total duration in milliseconds
     */
    getTotalStepDuration(steps) {
        return steps.reduce((sum, step) => sum + step.duration, 0);
    }
    /**
     * Get slowest step
     * @param steps - Array of step data
     * @returns Slowest step or null if no steps
     */
    getSlowestStep(steps) {
        if (steps.length === 0)
            return null;
        return steps.find(s => s.isSlowest) || null;
    }
}
exports.StepCollector = StepCollector;
